python/phase-1.md

```diff
@@ -0,0 +1,460 @@
+
+
+class MyActivities:
+    def activity1():
+        pass
+
+
+def create_my_worker(client: temporalio.Client) -> temporalio.Worker:
+    return temporalio.Worker(
+        client,
+        taskqueue="MyTaskQueue",
+        activities={
+            # The names can be any value
+            "simple_activity": simple_activity,
+            # Can use async function
+            "another_activity": another_activity,
+            # Can use callable class
+            "callable_class": MyCallableClass(),
+            # Can use method
+            "activity1_method": MyActivities().activity1,
+            # Can use lambda
+            "lambda": lambda arg1: arg1,
+            # Fails because not callable
+            "not_callable": MyActivities(),
+        },
+    )
+```
+
+Notes:
+
+* Cancellation is task cancellation
+  * We need to research whether, in Python 3.7, we can interrupt an async task w/ our own error reliably
+* Heartbeat is a synchronous call
+* There were a lot of discussions around registering of activities, here are the results:
+  * At least to start with, we want explicit names during register instead of the callable's `__name__` to avoid
+    name-registration surprises that plague other languages
+    * It was decided we can always be more flexible later or provided utilities if we must
+  * A decorator (e.g. `@activity.defn`) was gonna help with naming, but:
+    * Since we decided to use a mapping in the constructor instead of an iterable, name is always there
+    * So the decorator would have no current purpose except as a "marker" to implementers to beware how it is used,
+      which we decided was unnecessary at this time but we can revisit
+  * We still have to have discussions around activity invocation from a worker, because just what appears to be a normal
+    Python invocation turning to an `ExecuteActivity` or not based on decorator or similar alone may not be explicit
+    enough
+
+### Payload Converters
+
+Exported from `temporalio` (but maybe in `temporalio/converter/converter.py`, exported from `temporalio.converter`
+then exported from `temporalio`):
+
+```python
+class PayloadConverter(ABC):
+    # This can be sync or async. Always takes a list. Can return a single
+    # payload or a list of payloads with the same size as the input list or None
+    # if the value cannot be converted with this converter. Any other result
+    # causes an error. Any error raised is bubbled.
+    @abstractmethod
+    async def encode(
+        self, *values: Any
+    ) -> Union[temporalio.Payload, list[temporalio.Payload]]:
+        pass
+
+    # This can be sync or async. Always takes a list. Can return a single value
+    # or a list of values with the same size as the input list or None if the
+    # value cannot be converted with this converter. Any other result causes an
+    # error. Any error raised is bubbled.
+    @abstractmethod
+    async def decode(self, *payloads: temporalio.Payload) -> Union[Any, list[Any]]:
```

Comment:
There is an (obvious in hindsight) ambiguity issue here. What if your single payload is a list?

Any suggestions on how to combine "decode single payload -> multiple values" and "decode multiple payloads -> multiple values" and "decode multiple payloads -> single value"? Do we have to support the last use case?

I am now thinking about adding extra `encode_multiple` and `decode_multiple` calls that default to just calling these and if someone, say for encryption purposes, wants to convert multiple to single, they have to override those two also. Thoughts?

Reply:
Due to additional ambiguities, I'm gonna have to return a `Tuple[Any, bool]` from `decode(self, payload)` with the second value being `True` if it was valid. There is ambiguity if I just return `-> Any` because what if `None` is a real value?

Another approach would be to separate out into data and payload converters again, but that seems unnecesary.

Reply:
I thought these methods were meant for multiple payloads, it's okay to have a payload that's an array.
The method can map M to N payloads.